### PR TITLE
Add Semantics Property `linkUrl`

### PR DESCRIPTION
**Pull request by mdebbar**
The new property allows the user to specify a URI for their semantics link node. It's plumbed through for both web and non-web engines, but it's only used in the web engine currently. It sets the `href` of the anchor element associated with semantics node.

This is going to unlock better semantics support in the Link widget on web ([PR](https://github.com/flutter/packages/pull/6711)).

Framework counterpart: https://github.com/flutter/flutter/pull/150639

Part of https://github.com/flutter/flutter/issues/150263

### CODE REVIEW

File: lib/web_ui/lib/src/engine/semantics/semantics.dart [hunk 1]
```diff
@@ -337,6 +338,9 @@ class SemanticsNodeUpdate {
 
   /// See [ui.SemanticsUpdateBuilder.updateNode].
   final int headingLevel;
+
+  /// See [ui.SemanticsUpdateBuilder.updateNode].
+  final String? linkUri;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/semantics/semantics.dart:343]: `linkUri` is non-nullable in the `updateNode` API. Does it need to be nullable here?

File: lib/web_ui/lib/src/engine/semantics/link.dart [hunk 2]
```diff
@@ -18,9 +18,10 @@ class Link extends PrimaryRoleManager {
   @override
   DomElement createElement() {
     final DomElement element = domDocument.createElement('a');
-    // TODO(mdebbar): Fill in the real link once the framework sends entire uri.
-    // https://github.com/flutter/flutter/issues/150263.
     element.style.display = 'block';
+    if (semanticsObject.hasLinkUri) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/semantics/link.dart:22]: Can the link change? If yes, this code should probably go into an override of the `update` method and it should also check for `isLinkUriDirty`.

File: lib/ui/fixtures/ui_test.dart [hunk 3]
```diff
@@ -233,6 +233,7 @@ void sendSemanticsUpdate() {
     childrenInHitTestOrder: childrenInHitTestOrder,
     additionalActions: additionalActions,
     headingLevel: 0,
+    linkUri: '',
```
Comment (yjbanov) [inline lib/ui/fixtures/ui_test.dart:236]: I think "URL" is more accurate here than "URI". The `<a>` tag and the `Link` widget only work with URIs that are also fetchable over the network via the same URI. Using arbitrary URIs as abstract identifiers without being able to fetch them is not something that's supported by this property, AFAICT. Unless there's something on the mobile side that extends beyond that?

File: lib/web_ui/lib/src/engine/semantics/link.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/semantics/link.dart:22]: Good catch!

File: lib/web_ui/lib/src/engine/semantics/semantics.dart (hunk 1 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/semantics/semantics.dart:343]: This mirrors the code for the nullable `tooltip` field. I think it is non-nullable in the `updateNode` API to make it easier for the C++ side to handle.

File: lib/ui/fixtures/ui_test.dart (hunk 3 above)
Comment (harryterkelsen) [reply lib/ui/fixtures/ui_test.dart:236]: Done

File: lib/web_ui/lib/src/engine/semantics/link.dart (hunk 2 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/semantics/link.dart:29]: Done

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**